### ISSUE TITLE
Identifying Datasets with Missing Owners

### ISSUE THREAD

**Issue by teeceeas** [ministryofjustice/find-moj-data#910]

**Comment by MatMoore** [comment]
There are some notes on https://github.com/ministryofjustice/find-moj-data/issues/660 for running SQL against the datahub db. Make sure we don't run against the primary in production though. Do we have a read replica that we could point a BI tool to?

Potentially speak to AP to see whether Quicksight on AP could be used for this.

**Comment by MatMoore** [comment]
[Example of setting up a read replica on cloud platform](https://github.com/ministryofjustice/cloud-platform-environments/blob/b51314640a4413b95b8f3837131b1df44d2aba64/namespaces/live.cloud-platform.service.justice.gov.uk/laa-assess-crime-forms-dev/resources/rds-postgresql.tf#L41)

**Comment by MatMoore** [comment]
Some options for surfacing missing metadata

## 1. In Find MoJ Data, or another bespoke app

## 2. In AP, via Quicksight
TODO: find out more

## 3. In Power BI
We can access the Power BI web interface & share reports via Microsoft 365. However, we can't connect this to arbitrary data sources through the web interface, and the desktop version only runs on windows. So I don't think we should use this, given we are using macs.

## 3. In an open source reporting tool we self-host
E.g. redash, or Apache Superset

This seems overkill for what we need at the moment, as we would have to learn & maintain a new product, so I'm ruling this out *for now*

**Comment by MatMoore** [comment]
## Metadata completeness framework

This spreadsheet contains the fields that:

- are displayed in Find MoJ data
- are not automatically populated (e.g. column types)

https://docs.google.com/spreadsheets/d/1wqSFTp2_BrDxDuS7dfH6TdLmgv1GWvovy7kn0lEpjWc/edit?usp=sharing

I recommend starting out by focusing on owner, subject area and descriptions.

- The most critical is that some kind of owner is set - without it we have no accountability and reporting on the other metadata items is unlikely to get us anywhere.
- Subject area is important as it helps users find data that is relevant to them
- Description is important as it helps user decide whether the data they have found is relevant or not

**Comment by jemnery** [comment]
While we're in the PoC / R&D stage it may be wise to avoid #1 ("In Find MoJ Data, or another bespoke app") until we have a much clearer idea of what the reporting requirements are. Even if that is the most desirable end state for surfacing issues and success stories to our users.

Otherwise we risk building things which don't end up being useful.

**Comment by murdo-moj** [comment]
On the spreadsheet, it would also be useful to have breakdowns for populations of owner, subject area, and descriptions on other fields so data owners can see "their" population rates at a glance. A good place to start would be to break the stats down per database/dashboard/container.

**Comment by MatMoore** [comment]
@murdo-moj yeah agreed, I've added a list of possible breakdowns to the 2nd tab of the spreadsheet.

I was originally thinking we would break it down by owner, but perhaps Platform/Dashboard/Container is a better starting point, while we have incomplete owners.

**Comment by MatMoore** [comment]
TODO:

If we can get access to Quicksight, try it out with some fake metadata.

E.g.

- For a particular platform/container, show the % completeness of a a field (e.g. table description), aggregated across all tables
- For a particular platform/container, show the % completeness of a field (e.g. column description), aggregated across all columns and tables
- Present all tables/columns with missing field X, filtered by platform/container

**Comment by MatMoore** [comment]
Spoke to Darren and Julia and the AP Quicksight can definitely support this. For the purposes of the spike I will be able to test in data engineering sandbox A.

Quicksight will integrate nicely with Find MoJ data as it's possible to embed the dashboard into another service. AP can configure it so it's whitelisted to the Find MoJ data domain, and then we will serve it on a new page, behind authentication.

In terms of getting data in, we discussed 2 possible options:

Option 1 is making a direct connection from Quicksight to Datahub's postgres database. I'll need to check the networking is possible between modernisation platform and cloud platform, but Julia thinks it is.

Option 2 is creating an S3 bucket in AP, in the usual way (@LavMatt and @murdo-moj might need to help with this). Then we would schedule a service to run within the Datahub namespace and extract data to the bucket. It is possible to just dump arbitrary data here, but ideally, we would output the data in the a format that athena can read (maybe parquet files?), because quicksight integrates well with athena.

**Comment by MatMoore** [comment]
Things to test (TODO: add some screenshots)

- [x] Group by container urn to show the total number of missing values (e.g. for the description field)
- [ ] Show percent complete instead of sum
- [x] Filter a dataset
- [ ] Add controls (e.g. to choose which fields are empty)
- [ ] Lock down a published dashboard by domain (for embedding)
- [x] "Calculated fields" - this has a confusing syntax, so I would recommend modifying the dataset query rather than using this

**Comment by MatMoore** [comment]
## Granting access through Quicksight itself
https://docs.aws.amazon.com/quicksight/latest/user/share-a-dashboard-grant-access-everyone.html

## Embedding dashboards in Find MoJ data
It is possible to embed dashboards on a page in the Find MoJ data app.

See https://docs.aws.amazon.com/quicksight/latest/user/embedded-analytics-dashboards-for-everyone.html

AllowedDomains should be set to our Find MoJ data domains.
